### PR TITLE
Dispose response content on retry for unbuffered streams

### LINKED ISSUE #13546: Connections is being leaked when non-buffered response is retried on .NET Framework
Because we don't dispose response stream in retry policy.

@kasobol-msft, FYI this affects blob download.

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/13546

### CODE REVIEW

File: sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs [hunk 1]
```diff
@@ -127,6 +127,9 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                     }
                 }
 
+                // Dispose the content stream to free up a connection if the request has any
+                message.Response.ContentStream?.Dispose();
```
Comment (kasobol-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs:131]: Should this also happen in line 110 and/or 99 when retry loop exists with failure?

Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs]: At that point the stream become the responsibility of the caller. I.e. the logic that creates an exception would read and dispose the stream.

We just need to make sure that no streams are lost in retry process.

**Review by kasobol-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs [hunk 2]
```diff
@@ -127,6 +127,12 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                     }
                 }
 
+                if (message.HasResponse)
+                {
+                    // Dispose the content stream to free up a connection if the request has any
+                    message.Response.ContentStream?.Dispose();
```
Comment (KrzysztofCwalina) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs:133]: Can't response be null for some custom transports?

Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs:133]: No, response has to be set after transport returned from Process[Async] method. This is one of the core assumptions of our pipeline.

File: sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs [hunk 3]
```diff
@@ -81,6 +82,63 @@ public async Task NonBufferedExtractedStreamReadableAfterMessageDisposed()
...
+                    await httpPipeline.SendAsync(message, CancellationToken.None);
+
+                    Assert.AreEqual(message.Response.ContentStream.CanSeek, false);
+
+                    extractedStream = message.ExtractResponseContent();
+                }
+
+                var memoryStream = new MemoryStream();
+                await extractedStream.CopyToAsync(memoryStream);
+                Assert.AreEqual(memoryStream.Length, 1000);
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs:135]: nit: extract `1000` as a local and use it in `TestServer` as well.

File: sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs [hunk 4]
```diff
@@ -81,6 +82,63 @@ public async Task NonBufferedExtractedStreamReadableAfterMessageDisposed()
...
+                extractedStream.Dispose();
+            }
+
+            Assert.Greater(reqNum, 100);
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs:139]: nit: same here

**Review by christothes — APPROVED** [review]

**Pull request merged**